### PR TITLE
feat(connector-runtime): allow enabling/disabling of specific connectors

### LINKED ISSUE #1083: FR: ability to enable/disable specific connectors
__Is your feature request related to a problem? Please describe.__

There is a need to have an ability to turn on/off specific connectors in given runtime.

__Describe the solution you'd like__

**Option 1: connectors block list.** By default, all connectors are enabled. To disable, a user has to pass a flag to disable a connector.

**Option 2: connectors allow list.** By default, all connectors are enabled. To enable, a user has to pass a flag to disable a connector.

Similar to the [PR](https://github.com/camunda/connectors/pull/1068), flags can be environment variables, e.g: `CONNECTOR_HTTP_REST_ENABLED=true` or similar.

__Describe alternatives you've considered__

__Additional context__

Following up from the architecture review session 17 (Aug. 29, 2023).

### PR DESCRIPTION
**Pull request by vringar**
## Related issues

closes #1083

## Checklist

- [x] PR has a **milestone** or the `no milestone` label.

### CODE REVIEW

**Comment by vringar** [comment]
@chillleader This approach does not account for Spring Beans as they get added later. Is this something we need to consider here?

**Comment by maff** [comment]
> This approach does not account for Spring Beans as they get added later. Is this something we need to consider here?

Is Agentic AI the only solution using Spring Beans at the moment? We currently expose config properties to disable registration for individual connectors, so maybe that could be a general guideline for connectors registered via Spring Beans. For example, the ad-hoc tools schema resolver connector is [only registered](https://github.com/camunda/connectors/blob/main/connectors/agentic-ai/src/main/java/io/camunda/connector/agenticai/autoconfigure/AgenticAiConnectorsAutoConfiguration.java#L105) when `camunda.connector.agenticai.ad-hoc-tools-schema-resolver.enabled` is set to `true`.

What would be a great addition to this change would be to add a way to disable the default env/SPI discovery completely with a configuration value. Right now, it will always load connectors [via SPI unless a connector is registered via env variable](https://github.com/camunda/connectors/blob/main/connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/outbound/OutboundConnectorDiscovery.java#L27) (same for inbound). An additional configuration to disable both methods would be great for cases where only a set of Spring Bean connectors should be loaded.

**Comment by vringar** [comment]
Okay, so what I have rn is the following:

CONNECTOR_INBOUND/OUTBOUND_DISCOVERY_DISABLED: Removes all SPI/Env Variable lookup
CONNECTOR_INBOUND/OUTBOUND_DISABLES: Expects a comma separated list of connector types. This filters them out from the discovery and **prevents them from being added via registerConfiguration**.

@maff I chose to include this durable disabling because we want to support customer using connectors outside of spring, so there needs to be some disabling mechanism in core.

File: connector-runtime/connector-runtime-spring/src/main/java/io/camunda/connector/runtime/outbound/OutboundConnectorRuntimeConfiguration.java [hunk 1]
```diff
@@ -43,8 +44,10 @@ public class OutboundConnectorRuntimeConfiguration {
 
   @Bean
   public OutboundConnectorFactory outboundConnectorFactory() {
+    // filter out connectors that are disabled via additional env variable
     return new DefaultOutboundConnectorFactory(
-        OutboundConnectorDiscovery.loadConnectorConfigurations());
+        OutboundConnectorDiscovery.loadConnectorConfigurations(),
+        EnvVarsConnectorDisabler.getDisabledOutboundConnectorTypes());
```
Comment (sbuettner) [inline connector-runtime/connector-runtime-spring/src/main/java/io/camunda/connector/runtime/outbound/OutboundConnectorRuntimeConfiguration.java:50]: Curios: Wouldnt it be simpler for filter the list of configs before we handover the list to the factory so they are not even considered?

File: connector-runtime/spring-boot-starter-camunda-connectors/src/test/java/io/camunda/connector/runtime/app/EnvVarContextInitializer.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+import java.lang.reflect.InvocationTargetException;
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.env.EnvironmentPostProcessor;
+import org.springframework.core.env.ConfigurableEnvironment;
+
+/**
+ * This is a hack to inject the "environment variables" before the Configurations are loaded, so
+ * that the env vars are considered during construction of the inbound and outbound factories.
+ */
```
Comment (sbuettner) [inline connector-runtime/spring-boot-starter-camunda-connectors/src/test/java/io/camunda/connector/runtime/app/EnvVarContextInitializer.java:27]: I dont fully understand this comment: 
> hat the env vars are considered during construction of the inbound and outbound factories.

Cant we model the "filters" by having dedicated beans that provide this kind of functionality so they can easily be mocked in the tests?

File: connector-runtime/connector-runtime-spring/src/main/java/io/camunda/connector/runtime/outbound/OutboundConnectorRuntimeConfiguration.java (hunk 1 above)
Comment (vringar) [reply connector-runtime/connector-runtime-spring/src/main/java/io/camunda/connector/runtime/outbound/OutboundConnectorRuntimeConfiguration.java:50]: I wanted to do this initially, but I think if a connector is disabled, I believe you shouldn't be able to add it as well. So the list of disabled connector to be preserved.

File: connector-runtime/spring-boot-starter-camunda-connectors/src/test/java/io/camunda/connector/runtime/app/EnvVarContextInitializer.java (hunk 2 above)
Comment (vringar) [reply connector-runtime/spring-boot-starter-camunda-connectors/src/test/java/io/camunda/connector/runtime/app/EnvVarContextInitializer.java:27]: That's a very elegant idea. This would allow for a more powerful features down the line

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/EnvironmentVariablesAdapter.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * limitations under the License.
+ */
+package io.camunda.connector.runtime.core.discovery;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class EnvironmentVariablesAdapter {
```
[the remark this answers is not in the document]
Comment (vringar) [reply connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/EnvironmentVariablesAdapter.java:24]: Yes, and I hope we can eliminate it during the refactor to the `withEnvironment` approach

**Comment by vringar** [comment]
@sbuettner and I agreed that there is a lot of potential to clean this up, e.g. by removing the ability to dynamically register and just reordering the bean construction order to first collect all configurations (both Spring and SPI/Env), then apply the postprocessing steps of renaming and then the filtering, before ever construction the registry.

However, this PR aims to provide the minimal implementation and testing setup for disabling via env variables and leaves the rest as future work for the hacking session

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/EnvironmentVariablesAdapter.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.camunda.connector.runtime.core.discovery;
+
+import java.util.Optional;
+
+public class EnvironmentVariablesAdapter {
```
Comment (johnBgood) [inline connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/EnvironmentVariablesAdapter.java:21]: we could specify the class name and maybe even the method name. It was not clear to me what it did just looking at the method

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/EnvVarsConnectorDisabler.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+ */
+package io.camunda.connector.runtime.core.discovery;
+
+import static io.camunda.connector.runtime.core.discovery.EnvironmentVariablesAdapter.getEnv;
+
+import java.util.Arrays;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class EnvVarsConnectorDisabler {
```
Comment (johnBgood) [inline connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/EnvVarsConnectorDisabler.java:25]: not sure about this name again, one might think it actually disables something

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/EnvVarsConnectorDisabler.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+  public static boolean isOutboundDiscoveryDisabled() {
+    return getEnv("OUTBOUND", "DISCOVERY_DISABLED").isPresent();
+  }
+
+  public static Set<String> getDisabledOutboundConnectorTypes() {
+    return getDisabledConnectorTypes("OUTBOUND");
+  }
+
+  private static Set<String> getDisabledConnectorTypes(String direction) {
```
Comment (johnBgood) [inline connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/EnvVarsConnectorDisabler.java:43]: `direction` ?

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/inbound/DefaultInboundConnectorFactory.java [hunk 7]
```diff
@@ -99,10 +108,25 @@ public void resetConfigurations() {
...
+      return;
+    }
     if (EnvVarsConnectorDiscovery.isInboundConfigured()) {
       configurations = EnvVarsConnectorDiscovery.discoverInbound();
     } else {
       configurations = SPIConnectorDiscovery.discoverInbound();
     }
+    configurations =
+        configurations.stream()
+            .filter(
```
Comment (johnBgood) [inline connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/inbound/DefaultInboundConnectorFactory.java:122]: nit: for readability maybe we can extract the filter:
```
.filter(keepNotDisabled()) // a better name can be found :D
...
```

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/outbound/DefaultOutboundConnectorFactory.java [hunk 8]
```diff
@@ -38,9 +34,24 @@ public class DefaultOutboundConnectorFactory implements OutboundConnectorFactory
...
+      "Connector {} has been disabled by the CONNECTORS_INBOUND_DISABLED environment variable";
+
+  private final Set<String> disabledConnectors;
+
+  public DefaultOutboundConnectorFactory(
+      List<OutboundConnectorConfiguration> configurations, Set<String> disabledOutboundConnectors) {
+    this.disabledConnectors = disabledOutboundConnectors;
     connectorConfigs =
         configurations.stream()
+            .filter(
```
Comment (johnBgood) [inline connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/outbound/DefaultOutboundConnectorFactory.java:47]: same

**Review by johnBgood — DISMISSED** [review]
Left a few nitpicks, great stuff!

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/EnvVarsConnectorDisabler.java (hunk 6 above)
Comment (johnBgood) [reply connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/EnvVarsConnectorDisabler.java:43]: haaaa ok, I worked in the Search API a bit too much :D direction=sort direction to me

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/DisabledConnectorEnvVarsConfig.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    return getConnectorEnvironmentVariable(direction.name(), "DISCOVERY_DISABLED").isPresent();
+  }
+
+  public static boolean isConnectorDisabled(ConnectorConfiguration config) {
+    var direction =
+        switch (config) {
+          case OutboundConnectorConfiguration outboundConfig -> ConnectorDirection.OUTBOUND;
+          case InboundConnectorConfiguration inboundConfig -> ConnectorDirection.INBOUND;
+        };
+    var isDisabled = isConnectorDisabled(direction, config.type().toLowerCase());
```
Comment (johnBgood) [inline connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/DisabledConnectorEnvVarsConfig.java:43]: config.direction()

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/DisabledConnectorEnvVarsConfig.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+      LOG.info(
+          "Connector {} has been disabled by the CONNECTORS_{}_DISABLED environment variable",
+          config.type(),
+          direction.name());
+    }
+    return isDisabled;
+  }
+
+  private static boolean isConnectorDisabled(
+      ConnectorDirection connectorDirection, String lowerCase) {
```
Comment (johnBgood) [inline connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/DisabledConnectorEnvVarsConfig.java:54]: second param should be `type`

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/EnvVarUtil.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.camunda.connector.runtime.core.discovery;
+
+import java.util.Optional;
+
+public class EnvVarUtil {
```
Comment (johnBgood) [inline connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/EnvVarUtil.java:21]: `ConnectorEnvVars` ?

**Review by johnBgood — APPROVED** [review]
Left a few comments, great job on this PR!

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/DisabledConnectorEnvVarsConfig.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    return getDisabledConnectorTypes(connectorDirection).contains(lowerCase);
+  }
+
+  private static Set<String> getDisabledConnectorTypes(ConnectorDirection direction) {
+    return getConnectorEnvironmentVariable(direction.name(), "DISABLED")
```
[the remark this answers is not in the document]
Comment (vringar) [reply connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/DisabledConnectorEnvVarsConfig.java]: Okay, I believe the instance approach works better.

File: connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/DisabledConnectorEnvVarsConfig.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+import org.slf4j.LoggerFactory;
+
+public class DisabledConnectorEnvVarsConfig {
+  private static final Logger LOG = LoggerFactory.getLogger(DisabledConnectorEnvVarsConfig.class);
+
+  public static boolean isDiscoveryDisabled(ConnectorDirection direction) {
+    return getConnectorEnvironmentVariable(direction.name(), "DISCOVERY_DISABLED").isPresent();
+  }
+
+  public static boolean isConnectorDisabled(ConnectorConfiguration config) {
```
[the remark this answers is not in the document]
Comment (johnBgood) [reply connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/discovery/DisabledConnectorEnvVarsConfig.java]: that's probably enough then, just wanted to double-check 👍

File: connector-runtime/spring-boot-starter-camunda-connectors/src/test/java/io/camunda/connector/runtime/outbound/RuntimeStartupWithConnectorsFromEnvVarsTests.java [hunk 14]
```diff
@@ -51,18 +48,18 @@ class RuntimeStartupWithConnectorsFromEnvVarsTests {
...
+  //    EnvVarsConnectorDiscovery.clearHardwiredEnvironmentVariable();
+  //  }
+  //
+  //  @BeforeTestClass
+  //  public void prepare() {
+  //    EnvVarsConnectorDiscovery.addHardwiredEnvironmentVariable(
+  //        "CONNECTOR_TEST2_FUNCTION", "io.camunda.connector.http.HttpJsonFunction");
+  //    EnvVarsConnectorDiscovery.addHardwiredEnvironmentVariable(
+  //        "CONNECTOR_TEST2_TYPE", "non-default-TEST-task-type");
+  //  }
```
[the remark this answers is not in the document]
Comment (vringar) [reply connector-runtime/spring-boot-starter-camunda-connectors/src/test/java/io/camunda/connector/runtime/outbound/RuntimeStartupWithConnectorsFromEnvVarsTests.java:62]: @chillleader We previously kept this around in a disabled state, now it's even more disabled. Do we just rip it and try to reimplement it when we figure out how?

**Comment by johnBgood** [comment]
# QA information
## Test Environment
- Test required in (you can check both):
  - [ ] SaaS
  - [ ] SM
- [ ] Create a new test for this feature in our [Pre Release Tests](https://modeler.ultrawombat.com/projects/017779a8-9459-4feb-8b47-eeef8b2717ef--connectors-pre-release-test) for @Szik to be able to test

## Test Scope
*Please describe the test scope, happy path, edge cases that come to your mind, and whatever you think might relevant to test*

## Test Data
*Please provide the test data, if needed (files, URLs, code snippets, FEEL expressions)*

**Comment by johnBgood** [comment]
Could you create a follow-up ticket to implement the `ENABLE_...` equivalent of this feature? That would allow customers to either DISABLE some connectors, or ENABLE some (depending on the number of connectors).

**Comment by vringar** [comment]
Issue created https://github.com/camunda/connectors/issues/5177

**Review by Szik — APPROVED** [review]

**Review by johnBgood — APPROVED** [review]
